lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+       is available), we should produce the fallback code from
+       [direct].
+
+      (Note: [direct] is also a function (on [unit]) to ensure that any
+      effects performed during code production will only happen once we
+      do know that we want to produce the direct-style code.)
+
+     - [has_tmc_calls] is true when there are TMC opportunities
+       in the subterm -- if some calls are in tail-modulo-cons
+       position and are rewritten into tailcalls in the [dps] version.
+
+     - [benefits_from_dps] is true when the [dps] calls strictly more
+       TMC functions than the [direct] version. See the
+       {!choice_makeblock} case.
+
+     - [explicit_tailcall_request] is true when the user
+       used a [@tailcall] annotation on the optimizable callsite.
+       When one of several calls could be optimized, we expect that
+       exactly one of them will be annotated by the user, or fail
+       because the situation is ambiguous.
+   *)
+
+  let return (v : lambda) : lambda t = {
+    dps = Dps.return v;
+    direct = (fun () -> v);
+    has_tmc_calls = false;
+    benefits_from_dps = false;
+    explicit_tailcall_request = false;
+  }
+
+  let map f s = {
+    dps = Dps.map f s.dps;
+    direct = (fun () -> f (s.direct ()));
+    has_tmc_calls = s.has_tmc_calls;
+    benefits_from_dps = s.benefits_from_dps;
+    explicit_tailcall_request = s.explicit_tailcall_request;
+  }
+  (** Apply function [f] to the transformed term. *)
+
+  let direct (c : 'a t) : 'a =
+    c.direct ()
+
+  let dps (c : lambda t) ~tail ~dst =
+    Dps.run c.dps ~tail ~dst
+
+  let pair ((c1, c2) : 'a t * 'b t) : ('a * 'b) t = {
+    dps = Dps.pair c1.dps c2.dps;
+    direct = (fun () -> (c1.direct (), c2.direct ()));
+    has_tmc_calls =
+      c1.has_tmc_calls || c2.has_tmc_calls;
+    benefits_from_dps =
+      c1.benefits_from_dps || c2.benefits_from_dps;
+    explicit_tailcall_request =
+      c1.explicit_tailcall_request || c2.explicit_tailcall_request;
+  }
+
+  let unit = {
+    dps = Dps.unit;
+    direct = (fun () -> ());
+    has_tmc_calls = false;
+    benefits_from_dps = false;
+    explicit_tailcall_request = false;
+  }
+
+  module Syntax = struct
+    let (let+) a f = map f a
+    let (and+) a1 a2 = pair (a1, a2)
+  end
+  open Syntax
+
+  let option (c : 'a t option) : 'a option t =
+    match c with
+    | None -> let+ () = unit in None
+    | Some c -> let+ v = c in Some v
+
+  let rec list (c : 'a t list) : 'a list t =
+    match c with
+    | [] -> let+ () = unit in []
+    | c :: cs ->
+        let+ v = c
+        and+ vs = list cs
+        in v :: vs
+
+  (** The [find_*] machinery is used to locate a single subterm
+      to optimize among a list of subterms. If there are several possible choices,
+      we require that exactly one of them be annotated with [@tailcall], or
+      we report an ambiguity. *)
+  type 'a tmc_call_search =
+    | No_tmc_call of 'a list
+    | Nonambiguous of 'a zipper
+    | Ambiguous of 'a t list
```

Comment:
I don't think that there is a point in having that type parametrised, it suggest that it might be used in ways that never occurs. Same applies for zipper